### PR TITLE
SmallVector iteration improvements

### PR DESCRIPTION
**Pull request by kripken**
I needed these to experiment with replacing the vector in CFGWalker. However, I saw
no benefit to doing so, so I won't be landing that. But these iterator improvements
may help in the future (and I had to fight the C++ compiler quite a bit for them...).

### CODE REVIEW

File: src/support/small_vector.h [hunk 1]
```diff
@@ -162,6 +164,7 @@ template<typename T, size_t N> class SmallVector {
     }
 
     void operator++() { index++; }
+    void operator++(int) { index++; }
```
Comment (tlively) [inline src/support/small_vector.h:167]: These operators need to return `Iterator&` and `Iterator`. See https://github.com/WebAssembly/binaryen/blob/main/src/support/parent_index_iterator.h#L58-L75 for reference.

File: src/support/small_vector.h [hunk 2]
```diff
@@ -148,9 +148,11 @@ template<typename T, size_t N> class SmallVector {
   // iteration
 
   template<typename Parent, typename Iterator> struct IteratorBase {
+    using iterator_category = std::random_access_iterator_tag;
```
Comment (tlively) [inline src/support/small_vector.h:151]: It looks like there are still a bunch of operations missing to make this a valid random access iterator. See https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator for the docs.

Comment (kripken) [reply src/support/small_vector.h:153]: Thanks, I added a TODO for this.

File: src/support/small_vector.h (hunk 1 above)
Comment (kripken) [reply src/support/small_vector.h]: Added.

File: src/support/small_vector.h [hunk 3]
```diff
@@ -161,7 +165,14 @@ template<typename T, size_t N> class SmallVector {
...
 
-    void operator++() { index++; }
+    Iterator& operator++() {
+      index++;
+      return *this;
+    }
+    Iterator operator++(int) {
+      index++;
+      return *this;
+    }
```
Comment (tlively) [inline src/support/small_vector.h:175]: We should at least make this a correct postincrement implementation.
```suggestion
    Iterator operator++(int) {
      Iterator self = *static_cast<Iterator*>(this);
      index++;
      return self;
    }
```

**Review by tlively — APPROVED** [review]
I would be more comfortable if we implemented the missing functionality, but also doing that lazily once the compiler complains doesn't seem like the end of the world.

**Comment by nunotexbsd** [comment]
@kripken 

Could you please take a look at https://github.com/WebAssembly/binaryen/issues/5928 ?
Thanks

**Pull request merged**